src/agent/src/device/mod.rs

```diff
@@ -238,6 +244,49 @@ pub async fn add_devices(
     update_spec_devices(logger, spec, dev_updates)
 }
 
+#[instrument]
+pub async fn handle_cdi_devices(
+    logger: &Logger,
+    devices: &[Device],
+    spec: &mut Spec,
+    sandbox: &Arc<Mutex<Sandbox>>,
+) -> Result<()> {
+    if let Some(container_type) = spec
+        .annotations()
+        .as_ref()
+        .unwrap()
+        .get("io.katacontainers.pkg.oci.container_type")
+    {
+        if container_type == "pod_sandbox" {
+            return Ok(());
+        }
+    }
+
+    let (_keys, devices) = parse_annotations(&spec.annotations().as_ref().unwrap())?;
+
+    if devices.is_empty() {
+        info!(logger, "no CDI annotations, no devices to inject");
+        return Ok(());
+    }
+
+    for _ in 0..=CDI_TIMEOUT_LIMIT {
+        match inject_devices(spec, devices.clone()) {
+            Ok(_) => {
+                info!(logger, "all devices injected successfully");
+                return Ok(());
+            }
+            Err(e) => {
+                info!(logger, "error injecting devices: {:?}", e);
```

Comment:
Do you think we can skip the loop if we get an error and return the result?

Reply:
There's a timeout CDI_TIMEOUT_LIMIT, it will do its best to inject devices till it exceeds CDI_TIMEOUT_LIMIT.
In this aspect, I think it's fine to do so.

Reply:
We cannot just return since we do not know when the hot-plug event and CDI spec creation are complete.